text/0143-sdk-fail-safe-mode.md

```diff
@@ -0,0 +1,512 @@
+
+### Option A3: Stacktrace Detection <a name="option-a3"></a>
+
+Before sending a crash report, the SDK identifies an SDK crash by looking at the topmost frames of the crashing thread. If the topmost frames stem from the SDK itself, it disables itself. The [SDK crash detection](https://github.com/getsentry/sentry/tree/master/src/sentry/utils/sdk_crashes) already uses this approach in the event processing pipeline.
+
+#### Continuous Crashing Scenarios <a name="option-a3-continuous-crashing-scenarios"></a>
+
+Notes on [continuous crashing scenarios](#continuous-crash-scenarios):
+
+| Scenario | Covered | Notes |
+| --- | --- | --- |
+| 1. Worst Case - SDK Continuously Crashing During App Start No Crash Reports | ⛔️ - no | The stacktrace detection needs crash reports to work. |
+| 2. Almost Worst Case - SDK Continuously Crashing During App Start Can Send Crash Reports | ✅ - yes |  |
+| 3. SDK Continuously Crashing Shortly After SDK Init | ✅ - yes |  |
+| 4. SDK Continuously Crashing After SDK Init | ✅ - yes |  |
+
+Notes on [potential false positives](#potential-false-positives):
+
+| Scenario | Covered | Notes |
+| --- | --- | --- |
+| 1. User's Application Crashes Before SDK Initialization | ⛔️ - no | No option can cover this. |
+| 2. User's Application Crashes Async During SDK Initialization | ✅ - yes | The stacktrace detection correctly ignores this scenario. |
+| 3. User's Application Crashes Shortly After SDK Initialization | ✅ - yes | The stacktrace detection correctly ignores this scenario. |
+| 4. User's Application Crashes After SDK Initialization | ✅ - yes | The stacktrace detection correctly ignores this scenario. |
+
+
+#### Pros <a name="option-a3-pros"></a>
+
+1. It requires little to no extra overhead.
+2. It can ignore async app crashes during SDK initialization.
+3. It is the most reliable option to detect if the SDK crashes.
+
+#### Cons <a name="option-a3-cons"></a>
+
+1. __Doesn't work with static linking:__ This approach doesn’t work with static linking, as the Sentry SDKs end up in the same binary as the main app. As we don’t have symbolication in release builds, we can’t reliably detect if the memory address stems from the Sentry SDK or the app. We might be able to compare addresses with known addresses of specific methods or classes, but this won’t work reliably. As with iOS, many apps use static linking, so we must use an alternative approach.
+2. __Doesn't work for obfuscated code:__ For obfuscated code, detecting if a frame in the stacktrace stems from the Sentry SDK or the app can be difficult or even impossible.
+3. __Wrongly disabling the SDK:__ We frequently see wrongly reported SDK crashes in the SDK crash detection. As SDKs use bytecode manipulation, swizzling, or monkey patching, the stacktraces sometimes contain Sentry frames in the crashing thread, but the root cause isn't Sentry but the user's code.
+4. It doesn't work when the SDK crashes during or before sending the crash report.
```

Comment:
just an idea crossed my mind: we could easily solve 4 and 5 cons here by installing another internal crash handler (if possible) as the first thing when initializing the SDK, whose sole purpose would be to identify SDK crashes and disable the SDK/let us know about/etc.

This way we could also checkmark the 1st scenario here and ultimately be the best solution? 😅 but the problem with static linking still remains ofc.

Maybe we could think of this as another potential solution or apply it to some other one.